src/ApiService/ApiService/onefuzzlib/ImageReference.cs

```diff
@@ -0,0 +1,247 @@
+            identifier = new ResourceIdentifier(image);
+            if (identifier.ResourceType == SharedGalleryImageResource.ResourceType) {
+                result = new LatestSharedGalleryImage(identifier);
+            } else if (identifier.ResourceType == SharedGalleryImageVersionResource.ResourceType) {
+                result = new SharedGalleryImage(identifier);
+            } else if (identifier.ResourceType == GalleryImageVersionResource.ResourceType) {
+                result = new GalleryImage(identifier);
+            } else if (identifier.ResourceType == GalleryImageResource.ResourceType) {
+                result = new LatestGalleryImage(identifier);
+            } else if (identifier.ResourceType == ImageResource.ResourceType) {
+                result = new Image(identifier);
+            } else {
+                return new Error(
+                    ErrorCode.INVALID_IMAGE,
+                    new[] { $"Unknown image resource type: {identifier.ResourceType}" });
+            }
+        } catch (FormatException) {
+            // not an ARM identifier, try to parse a marketplace image:
+            var imageParts = image.Split(":");
+            // The python code would throw if more than 4 parts are found in the split
+            if (imageParts.Length != 4) {
+                return new Error(
+                    Code: ErrorCode.INVALID_IMAGE,
+                    new[] { $"Expected 4 ':' separated parts in '{image}'" });
+            }
+
+            result = new Marketplace(
+                    Publisher: imageParts[0],
+                    Offer: imageParts[1],
+                    Sku: imageParts[2],
+                    Version: imageParts[3]);
+        }
+
+        return OneFuzzResult.Ok(result);
+    }
+
+    public abstract Task<OneFuzzResult<Os>> GetOs(ArmClient armClient, Region region);
+
+    public abstract Compute.Models.ImageReference ToArm();
+
+    public abstract long MaximumVmCount { get; }
+
+    // Documented here: https://docs.microsoft.com/en-us/azure/virtual-machine-scale-sets/virtual-machine-scale-sets-placement-groups#checklist-for-using-large-scale-sets
+    protected const long CustomImageMaximumVmCount = 600;
+    protected const long MarketplaceImageMaximumVmCount = 1000;
+
+    public abstract override string ToString();
+
+    public abstract record ArmImageReference(ResourceIdentifier Identifier) : ImageReference {
+        public sealed override long MaximumVmCount => CustomImageMaximumVmCount;
+
+        public sealed override Compute.Models.ImageReference ToArm()
+            => new() { Id = Identifier };
+
+        public sealed override string ToString() => Identifier.ToString();
+    }
+
+    [JsonConverter(typeof(Converter<LatestGalleryImage>))]
+    public sealed record LatestGalleryImage(ResourceIdentifier Identifier) : ArmImageReference(Identifier) {
+        public override async Task<OneFuzzResult<Os>> GetOs(ArmClient armClient, Region region) {
```

Comment:
In the previous `GetOs` implementation (in ImageOperations which is now deleted), we cached the request to Azure. We should make sure we continue caching for each of these new sub-implementations.

Reply:
Good call.